### PR TITLE
[DRAFT] Adds invitation_status to memberships

### PR DESCRIPTION
**Pull request by petrokoriakin1**

### CODE REVIEW

File: test/models/membership_test.rb [hunk 1]
```diff
@@ -34,4 +34,13 @@ class MembershipTest < ActiveSupport::TestCase
     assert_not membership.save
     assert_includes membership.errors.messages[:base], "Role cannot be changed because this is the only admin."
   end
+
```
Comment (petrokoriakin1) [inline test/models/membership_test.rb:37]: for some reason, the failure exists

```
% bin/rails test test/models/membership_test.rb
Running 4 tests in a single process (parallelization threshold is 50)
Run options: --seed 25789

# Running:

...E

Error:
MembershipTest#test_try_destroy:
NoMethodError: undefined method `pry' for an instance of Binding
    test/models/membership_test.rb:21:in `block in <class:MembershipTest>'


bin/rails test test/models/membership_test.rb:11



Finished in 0.202031s, 19.7989 runs/s, 64.3466 assertions/s.
4 runs, 13 assertions, 0 failures, 1 errors, 0 skips

```

File: .tool-versions [hunk 2]
```diff
@@ -0,0 +1 @@
+ruby 3.3.3
```
Comment (petrokoriakin1) [inline .tool-versions:1]: @yshmarov  are you OK with introducing `asdf` [tools version](https://asdf-vm.com/) control?

File: test/models/membership_test.rb (hunk 1 above)
Comment (petrokoriakin1) [reply test/models/membership_test.rb:37]: the resolution appeared to be simple

```ruby
  def admin_must_have_active_invitation_status
    return if role == "admin" && invitation_status == "active"
    return if role == "member" && invitation_status.present?

    errors.add(:invitation_status, "Admins must always have an active invitation status.")
  end
```

File: Gemfile [hunk 3]
```diff
@@ -52,6 +52,7 @@ group :development, :test do
   # Omakase Ruby styling [https://github.com/rails/rubocop-rails-omakase/]
   gem "rubocop-rails-omakase", require: false
   gem "erb_lint", require: false
+  gem "devpack"
```
Comment (petrokoriakin1) [inline Gemfile:55]: @bobf is this a good usage of [devpack](https://rubygems.org/gems/devpack)?

We try our best to keep the moneygun biolerplate LEAN and stick to defaults whenever we can.

File: .tool-versions (hunk 2 above)
Comment (yshmarov) [reply .tool-versions:1]: pls don't commit this. keep your dev environment config local

File: app/models/membership.rb [hunk 4]
```diff
@@ -19,6 +21,13 @@ def try_destroy
 
   private
 
+  def admin_must_have_active_invitation_status
```
Comment (yshmarov) [inline app/models/membership.rb:24]: what if I invite another admin, and he did not yet confirm the invitation?

Comment (petrokoriakin1) [reply app/models/membership.rb:24]: fixed now. thanks for catching that!

File: .tool-versions (hunk 2 above)
Comment (petrokoriakin1) [reply .tool-versions]: removing now.

File: app/models/membership.rb [hunk 5]
```diff
@@ -3,27 +3,40 @@ class Membership < ApplicationRecord
...
     return if organization.memberships.where(role: "admin").count > 1
 
     if role_changed? && role_was == "admin"
       errors.add(:base, "Role cannot be changed because this is the only admin.")
     end
   end
+
+  def single_admin_role_persent?
+    role == "admin" && organization.memberships.where(role: "admin").count == 1
+  end
```
Comment (petrokoriakin1) [inline app/models/membership.rb:41]: I'm not sure this is the responsibility of `Membership`... I'd rather try to make it `Organisation`'s attribute

Comment (yshmarov) [reply app/models/membership.rb:41]: why don't you build the actual invitation logic, and focus no such details later, while writing integration tests for possible edge cases?

**Comment by yshmarov** [comment]
closing as stale
